File: solutions/JavaScript/search_2d_matrix_binary_search.js
function searchMatrix(matrix, target) {
    if (!matrix || !matrix.length) return false;

    const m = matrix.length, n = matrix[0].length;
    let left = 0, right = m * n - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const mid_value = matrix[Math.floor(mid / n)][mid % n];

        if (mid_value === target) return true;
        else if (mid_value < target) left = mid + 1;
        else right = mid - 1;
    }

    return false;
}

File: solutions/Golang/search_2d_matrix_binary_search.go
package main

func searchMatrix(matrix [][]int, target int) bool {
    if len(matrix) == 0 || len(matrix[0]) == 0 { return false }

    m, n := len(matrix), len(matrix[0])
    left, right := 0, m*n-1

    for left <= right {
        mid := (left + right) / 2
        midValue := matrix[mid/n][mid%n]

        if midValue == target {
            return true
        } else if midValue < target {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return false
}
